### ISSUE TITLE
[question] Maskable PPO: creating an actions mask for multibinary action space ISSUE 

### ISSUE THREAD

**Issue by aliirshayyid** [Stable-Baselines-Team/stable-baselines3-contrib#102]
I'm trying to use MaskablePPO but I have a problem with the dimensions of the mask ( what should be returned in env.valid_action_mask())! 
My costume environment action space is 
```python
self.action_space = spaces.MultiBinary(4) # 4 variables each has only two options
self.observation_space = spaces.Box(low=-500, high=255,
                                        shape=(16,), dtype=np.float64)
```
My env valid_action_mask is (basically, I take only the first four features of the observation, and if it's 1, then the corresponding action is invalid)
```python
    def valid_action_mask(self):
        obs = self.observation[0:4]
        mask =  np.array([s != 1 for s in obs]) # [True, False, True, True}
        return mask
```
This is the training code 
```python

def mask_fn(env: gym.Env) -> np.ndarray:
    # Do whatever you'd like in this function to return the action mask
    # for the current env. In this example, we assume the env has a
    # helpful method we can rely on.
    return env.valid_action_mask()

env = pltnMrgEnv()
env = ActionMasker(env, mask_fn)  # Wrap to enable masking
env.reset()

model = MaskablePPO(MaskableActorCriticPolicy, env, verbose=1)
model.learn(total_timesteps=100000)
```
And I get this error that doesn't make sense! 
```bash 
Traceback (most recent call last):
  File "c:\Users\ali\Merge_platoon_sim_vscode\trainWmaskableppo.py", line 49, in <module>
    model.learn(total_timesteps=100000)
  File "C:\Users\ali\anaconda3\lib\site-packages\sb3_contrib\ppo_mask\ppo_mask.py", line 559, in learn
    continue_training = self.collect_rollouts(self.env, callback, self.rollout_buffer, self.n_steps, use_masking)
  File "C:\Users\ali\anaconda3\lib\site-packages\sb3_contrib\ppo_mask\ppo_mask.py", line 330, in collect_rollouts
    actions, values, log_probs = self.policy(obs_tensor, action_masks=action_masks)
  File "C:\Users\ali\anaconda3\lib\site-packages\torch\nn\modules\module.py", line 1130, in _call_impl
    return forward_call(*input, **kwargs)
  File "C:\Users\ali\anaconda3\lib\site-packages\sb3_contrib\common\maskable\policies.py", line 115, in forward
    distribution.apply_masking(action_masks)
  File "C:\Users\ali\anaconda3\lib\site-packages\sb3_contrib\common\maskable\distributions.py", line 231, in apply_masking
    masks = masks.view(-1, sum(self.action_dims))
RuntimeError: shape '[-1, 8]' is invalid for input of size 4
```

I tracked the error and the problem is self.action_dims is [ 2,2,2,2], which makes the sum equal to 8, and the mask length is four elements only! 
Should I make the mask length 8? if so, what should the shape be? 
I'm assuming because it's binary, making the logits a large negative number makes the output always zero, which means the mask needs to be equal to the number of actions!! 
I'm not sure what I'm missing! Please help!
Thanks

**Comment by aliirshayyid** [comment]
Thank you so much for the quick reply!
I don't quite get it because my thinking is that False would mean don't fire that action which means it's zero but I changed the code to return 8 values in the mask as you suggested.
But I still got unexpected results! My expectation is if the mask is 8 Falses, this will make the actions all zeros, right? 

It's actually not! 
```python
def valid_action_mask(self):
        obs = self.observation[0:4]
        # mask =  np.array([[s != 1, s != 1] for s in obs])
        mask = np.array([[False, False], [False, False],[False, False],[False, False]])
        return mask
```
I just made the mask all False to check the effect! 
Yet I get [ 1, 0, 0, 0] action, and I don't understand why!

**Comment by araffin** [comment]
I think there is a misunderstanding of either the action space or the masking.
You have `spaces.MultiBinary(4) # 4 variables each has only two options`, so the agent **must** take 4 actions per step (each action is either 0 or 1), which means you can mask **at most** one action per dimension.

In the example you are providing, you are masking all actions, making all of them invalid, which lead to weird behavior.

**Comment by aliirshayyid** [comment]
Thank you so much for your help @araffin! 
That is correct. 
I feel it can be documented better, especially the mask part, so I will add my finding here, hoping it will help someone.
So basically, what you can tell from the documentation is that the mask should be a NumPy array of bool with the size of all possible actions (the mask's shape is not important).
For my case ( 4 binary inputs ), the mask should be something like this ``` [ [bool, bool], [bool, bool], [bool, bool],[ bool, bool]]```
Now the obvious question is which represents the "0" and which represents "1"?
The first element represents 0, and the second represents 1.  
In other words, 
```python
mask = np.array([True, False, True, False, True, False, True, False]) # yields all zeros,  actions = [0,0,0,0]
mask = np.array([ [False, True], [False, True], [False, True],[ False, True]]) # yields all ones, actions = [1,1,1,1]
```
Finally, this is my valid_action_mask function,
```python
def valid_action_mask(self):
        obs = self.observation[0:4]
        mask =  np.array([[True, s != 1] for s in obs]) # when s is one, only action 0 is valid( 1 is not valid)
        return mask
```
